ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ChainSync/Client/JumpingGovernor.hs

```diff
@@ -0,0 +1,337 @@
+    -- ClientState.
+    FoundIntersection
+  deriving (Generic)
+
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (JumperState blk)
+
+-- | Instruction from the jumping governor, either to run normal ChainSync, or
+-- to jump to follow the given peer with the given fragment.
+data Instruction blk
+  = RunNormally
+  | JumpTo !(Point blk)
+  deriving (Generic)
+
+deriving instance (HasHeader blk, Eq (Header blk)) => Eq (Instruction blk)
+deriving instance (HasHeader blk, Show (Header blk)) => Show (Instruction blk)
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (Instruction blk)
+
+data JumpResult blk
+  = AcceptedJump !(Point blk)
+  | RejectedJump !(Point blk)
+  deriving (Generic)
+
+deriving instance (HasHeader blk, Eq (Header blk)) => Eq (JumpResult blk)
+deriving instance (HasHeader blk, Show (Header blk)) => Show (JumpResult blk)
+deriving anyclass instance (HasHeader blk, NoThunks (Header blk)) => NoThunks (JumpResult blk)
+
+nextInstruction ::
+  ( MonadSTM m,
+    Ord peer,
+    HasHeader blk, HasHeader (Header blk)
+  ) =>
+  Handle m peer blk ->
+  peer ->
+  STM m (Instruction blk)
+nextInstruction governor@Handle {clientStates} peer =
+  ((Map.! peer) <$> readTVar clientStates) >>= \case
+    Dynamo lastJumpSlot -> maybeJump lastJumpSlot >> pure RunNormally
+    Objector _ -> pure RunNormally
+    Jumper nextJumpVar _ _ ->
+      readTVar nextJumpVar >>= \case
+        Nothing -> retry
+        Just fragment -> do
+          writeTVar nextJumpVar Nothing
+          pure $ JumpTo fragment
+  where
+    maybeJump lastJumpSlot = do
+      dynamoFragment <- fromJust <$> getDynamoFragment governor
+      when (succWithOrigin (headSlot dynamoFragment) >= succWithOrigin lastJumpSlot + jumpSize) $ do
+        clientStates' <- readTVar clientStates
+        forM_ (Map.elems clientStates') $ \case
+          Jumper nextJumpVar _ Happy -> writeTVar nextJumpVar $ Just $! castPoint (headPoint dynamoFragment)
+          _ -> pure ()
+
+processJumpResult ::
+  ( MonadSTM m,
+    Ord peer, HasHeader blk, HasHeader (Header blk)
+  ) =>
+  Handle m peer blk ->
+  peer ->
+  JumpResult blk ->
+  STM m ()
+processJumpResult governor@Handle {clientStates, clientHandles} peer jumpResult =
+  ((Map.! peer) <$> readTVar clientStates) >>= \case
+    Dynamo _ -> pure ()
+    Objector _ -> pure ()
+    Jumper nextJumpVar goodPoint state ->
+        case (state, jumpResult) of
+          (FoundIntersection, _) ->
+            -- Ignore jump results when we already found the intersection.
+            pure ()
+          (Happy, AcceptedJump goodPoint') -> do
+            -- The jump was accepted; we bump the peer's candidate fragment to
+            -- the dynamo's candidate fragment up to the accepted point.
+            dynamoFragment <- fromJust <$> getDynamoFragment governor
+            let slicedDynamoFragment = fst $ fromJust $ splitAfterPoint dynamoFragment goodPoint'
```

Comment:
Is `fromJust` going to provide the source location of the error? This was not the case with `head` in all compiler versions. Maybe `error` should be called explicitly here.